### PR TITLE
Add sparse tensor allreduce

### PR DESCRIPTION
**Pull request by pietern**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19443 Support sparse gradients in DistributedDataParallel&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15007365/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19146 Add sparse tensor allreduce**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14889547/)

Implemented only on ProcessGroupGloo, as an allgather of metadata
(sparse_dim, dense_dim, and nnz), followed by an allgather of indices,
followed by an allgather of values. Once these operations have
finished, all ranks locally compute a reduction over these sparse
tensors. Works for both CPU and CUDA tensors.

This surfaced a problem with the existing assumption of only modifying
tensors that are passed at the call site, because for sparse tensors
we don't know the dimensions of the output tensors before we run the
collective. To deal with this unknown, this commit adds a `result`
function to the `c10d::ProcessGroup::Work` class that returns a vector
of tensors.

It's a bit odd to have to retrieve the result through this function
only for operations on sparse tensors. To make this work irrespective
of tensor layout, we can create a follow-up commit to make all in
place operations make their results accessible through this function
as well. This doesn't break any existing contracts but does have the
potential to add interface ambiguity.

Differential Revision: [D14889547](https://our.internmc.facebook.com/intern/diff/D14889547/)

### CODE REVIEW

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 1]
```diff
@@ -547,6 +560,246 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
...
+          continue;
+        }
+        const auto actual = metadata[i].sizes();
+        AT_CHECK(actual == expected, "Sparse dimensions do not match");
+      }
+    }
+
+    // Gather all indices and all values.
+    auto indices = allgather_indices(input, metadata);
+    auto values = allgather_values(input, metadata);
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: should these two be done in one allgather?

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 2]
```diff
@@ -547,6 +560,246 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
...
+    return output.coalesce();
+  }
+
+  void run() override {
+    auto output = allreduce(inputs);
+
+    // Copy back to input tensors.
+    outputs.reserve(inputs.size());
+    for (size_t i = 0; i < inputs.size(); i++) {
+      outputs.push_back(output.clone());
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: Is this only for `all_reduce_multigpu` API? If yes, should we throw error instead if `inputs.size() > 1`? Because the docs says 

> The function operates in-place and requires that each tensor to be a GPU tensor on different GPUs.

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 3]
```diff
@@ -606,6 +859,74 @@ class AsyncAllreduceCUDAWork : public AsyncAllreduceWork {
...
+      auto indices = output.indices().to(
+          inputs[i].device(),
+          /*non_blocking=*/true,
+          /*copy=*/true);
+      auto values = output.values().to(
+          inputs[i].device(),
+          /*non_blocking=*/true,
+          /*copy=*/true);
+
+      outputs.push_back(
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: If we are not writing back to the input tensors for sparse use case, we should edit Python docs accordingly. BTW, we can still write to input tensors using `tensor.set_()`?

File: torch/csrc/distributed/c10d/init.cpp [hunk 4]
```diff
@@ -190,18 +219,24 @@ They are used in specifying strategies for reduction collectives, e.g.,
...
+              },
               py::arg("tensors"),
               py::arg("opts") = ::c10d::AllreduceOptions(),
               py::call_guard<py::gil_scoped_release>())
 
           .def(
               "allreduce",
               [](::c10d::ProcessGroup& pg,
-                 std::vector<at::Tensor>& xs,
+                 std::vector<at::Tensor> xs,
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/csrc/distributed/c10d/init.cpp]: It turns out pybind always converts Python containers to STL containers in place. Because of that, the pass by reference is useless. I only found out when working on this change, so changed it here. I'll change it back for this commit and update every call site in this file to take arguments by value in one big sweeping change instead of only changing these ones.

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 5]
```diff
@@ -160,8 +161,21 @@ void initializeStreamsEvents(
...
+    if (tensors[i].is_sparse()) {
+      if (tensors[i].is_coalesced()) {
+        c10::cuda::CUDACachingAllocator::recordStream(
+            tensors[i].indices().storage().data(), streams[i]);
+        c10::cuda::CUDACachingAllocator::recordStream(
+            tensors[i].values().storage().data(), streams[i]);
+      } else {
+        // We will need to coalesce first, which means new tensors will
+        // be allocated on the streams we just allocated, and there
+        // is no need to record them separately.
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: It won't be freed because the op holds on to the input tensors. The coalesce on this new stream will happen after any operations against the inputs because of `events[i].block()` that is run before this block.

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 6]
```diff
@@ -547,6 +560,246 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
...
+  const uint32_t tag;
+
+  // We share dimensionality about the sparse tensors before collecting
+  // their contents. We assume here that the maximum number of sparse
+  // and dense dimensions is 4. This is stored in a contiguous piece of
+  // memory so that we can easily run allgather on it.
+  //
+  // The layout of this memory is as follows:
+  //
+  //   - [0:4]: sparse dims
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: Sure. The reason I picked 4 here is that I never saw anything bigger and we need to have a constant maximum number that is shared between processes, or else require another collective before we can exchange these details.

File: torch/lib/c10d/ProcessGroupGloo.cpp [hunk 7]
```diff
@@ -547,6 +560,246 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
...
+  //   - [4:8]: dense dims
+  //   -   [8]: nnz
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: Yes.

File: torch/lib/c10d/ProcessGroupGloo.cpp (hunk 1 above)
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: This should be possible in theory, since we have all data. But the data types may be different (the indices are always `long` and the values may be anything). To combine them in a single call we'd need to cast both of them to `char` buffers since it would run on the byte level.

File: torch/lib/c10d/ProcessGroupGloo.cpp (hunk 2 above)
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: We should update and deprecate `all_reduce_multigpu` instead, IMO. We already have the multi-input collectives (for some at least), so I added this one here for parity with those. If we end up with a replicated DDP module with an embedding bag, the reducer will end up calling this multiple sparse inputs. This functionality needs to live somewhere and I rather have it here than move it back into the reducer where we'd compute a sum across replicas before passing it to the process group. Doing so would introduce an asymmetry with ProcessGroupNCCL, where multi input allreduce is valid (and will defer summing them to NCCL).

**Comment by pietern** [comment]
@mrshenli The variable to tensor unboxing is no longer needed now that there has been some progress with the tensor/variable merge. The update also removes some verbose copying code (separate copy calls for the indices and values) and replaces those with a straightforward `tensor.to()`.

**Review by mrshenli — APPROVED** [review]

**Comment by lironmo** [comment]
@pietern when is this excepted to be merge and what will be the rc version?

**Comment by pietern** [comment]
@lironmo This was merged in #22036 and you can use it today with the nightly builds.

**Comment by lironmo** [comment]
@pietern thanks, i tried it on gloo backend and it's works.

I have a few questions:
1.can i use it with nccl backend?
2.it's seems that the support of sparse tensors are low (mainly missing serialization - i converted to coo matrix and use pickle) - you are going to support torch.save?
3. probably not related to you but it's seems that there is a major change in the lr scheduler(related to the base lr, up until this version the  lr_scheduler.step(epoch_metrics) done before the optimizer.step(), but it's break the multi step scheduler...

**Comment by pietern** [comment]
Hi @lironmo ,

1. No, this is implemented only for the Gloo backend at this point in time. Added #22400 to track.
2. Can you create an issue for missing functionality? Then we'll be able to track and prioritize it.
3. This doesn't sound related to sparse allreduce. You still need an optimizer that works with sparse gradients, of course.

**Comment by lironmo** [comment]
thanks.
